include/ginkgo/core/base/math.hpp

```diff
@@ -1220,10 +1236,12 @@ GKO_INLINE GKO_ATTRIBUTES T safe_divide(T a, T b)
  * @return `true` if the value is NaN.
  */
 template <typename T>
-GKO_INLINE GKO_ATTRIBUTES std::enable_if_t<!is_complex_s<T>::value, bool>
-is_nan(const T& value)
+GKO_DEPRECATED("is_nan can't be used safely on the device (MSVC+CUDA)")
```

Comment:
I don't want to give that recommendation, since that misses most NaN values, this is only meant to recognize the value returned by `gko::nan`, while `is_nan` might also be useful in other cases. This is meant to discourage using the function at all.

Reply:
Ok, then explicitly state that this function is deprecated without replacement. Otherwise, it could lead to confusion.

Reply:
Do you really want to deprecate the whole function because one use case disappeared? To me, they fulfill two separate intentions because`is_nan_exact` can't detect all NaN values.
I would at least not mention the `is_nan_exact` alternative.

Reply:
It was the only use case, and the function itself cannot be used in CUDA device code, which makes it misplaced in this function